compiler/noirc_evaluator/src/ssa/opt/pure.rs

```diff
@@ -49,10 +49,23 @@ impl Ssa {
             function.dfg.set_function_purities(purities.clone());
         }
 
+        debug_assert!(purity_analysis_post_check(&self));
```

Comment:
We are cloning each purity above. We should be able to just do this debug assertion a single time before storing the purities in each function DFG